rules/ruleset.go

```diff
@@ -683,3 +685,524 @@ type int64Asc []int64
 func (a int64Asc) Len() int           { return len(a) }
 func (a int64Asc) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
 func (a int64Asc) Less(i, j int) bool { return a[i] < a[j] }
+
+// MutableRuleSet is an extention of a RuleSet that implements mutation functions.
+type MutableRuleSet interface {
+	RuleSet
+
+	// Schema returns the schema.Ruleset representation of this ruleset.
+	Schema() (*schema.RuleSet, error)
+
+	// Clone returns a copy of this MutableRuleSet
+	Clone() (MutableRuleSet, error)
+
+	// MarshalJSON serializes this RuleSet into JSON.
+	MarshalJSON() ([]byte, error)
+
+	// UnmarshalJSON deserializes this RuleSet from JSON.
+	UnmarshalJSON(data []byte) error
+
+	// AppendMappingRule creates a new mapping rule and adds it to this ruleset.
+	AddMappingRule(MappingRuleData) error
+
+	// UpdateMappingRule creates a new mapping rule and adds it to this ruleset.
+	UpdateMappingRule(MappingRuleUpdate) error
+
+	// DeleteMappingRule deletes a mapping rule
+	DeleteMappingRule(DeleteData) error
+
+	// AppendRollupRule creates a new rollup rule and adds it to this ruleset.
+	AddRollupRule(RollupRuleData) error
+
+	// UpdateRollupRule creates a new rollup rule and adds it to this ruleset.
+	UpdateRollupRule(RollupRuleUpdate) error
+
+	// DeleteRollupRule deletes a rollup rule
+	DeleteRollupRule(DeleteData) error
+
+	// Tombstone tombstones this ruleset and all of its rules.
+	Delete(UpdateMetadata) error
+
+	// Revive removes the tombstone from this ruleset. It does not revive any rules.
+	Revive(UpdateMetadata) error
+}
+
+// Schema returns the protobuf representation fo a ruleset
+func (rs ruleSet) Schema() (*schema.RuleSet, error) {
+	res := &schema.RuleSet{
+		Uuid:          rs.uuid,
+		Namespace:     string(rs.namespace),
+		CreatedAt:     rs.createdAtNanos,
+		LastUpdatedAt: rs.lastUpdatedAtNanos,
+		Tombstoned:    rs.tombstoned,
+		CutoverTime:   rs.cutoverNanos,
+	}
+
+	mappingRules := make([]*schema.MappingRule, len(rs.mappingRules))
+	for i, m := range rs.mappingRules {
+		mr, err := m.Schema()
+		if err != nil {
+			return nil, err
+		}
+		mappingRules[i] = mr
+	}
+	res.MappingRules = mappingRules
+
+	rollupRules := make([]*schema.RollupRule, len(rs.rollupRules))
+	for i, r := range rs.rollupRules {
+		rr, err := r.Schema()
+		if err != nil {
+			return nil, err
+		}
+		rollupRules[i] = rr
+	}
+	res.RollupRules = rollupRules
+
+	return res, nil
+}
+
+func (rs ruleSet) Clone() (MutableRuleSet, error) {
```

Comment:
Feels a bit dirty, would you mind doing it more properly in a followup diff that performs deep copy without going through the object->schema->object roundtrip?